### ISSUE TITLE
Edit Vector

### ISSUE THREAD

**Issue by baopmessi** [williamyang1991/StyleGANEX#20]
Thanks for awesome work!.
How can i obtain a attribute edit vector (exp: smile, glasses, ...)

**Comment by williamyang1991** [comment]
please refer to https://github.com/williamyang1991/StyleGANEX/issues/17

**Comment by baopmessi** [comment]
Thanks for answer quickly!
I load vector from interfaceGan and inference but it's not working. The face has no change as expected.
![image](https://github.com/williamyang1991/StyleGANEX/assets/34626618/ebedd2be-9391-49a5-af92-7fe47541296c)
![image](https://github.com/williamyang1991/StyleGANEX/assets/34626618/7da41dea-c6a3-42d9-870d-fb53a6d3511f)

**Comment by williamyang1991** [comment]
The vector used in our age model has been saved within the model ckpt:  `ckpt['editing_w']`

```python
def load_model(path, device):
    local_path = hf_hub_download('PKUWilliamYang/StyleGANEX', path)
    ckpt = torch.load(local_path, map_location='cpu')
    opts = ckpt['opts']
    opts['checkpoint_path'] = local_path
    opts['device'] = device
    opts = Namespace(**opts)
    pspex = pSp(opts).to(device).eval()
    pspex.latent_avg = pspex.latent_avg.to(device)
    if 'editing_w' in ckpt.keys():
        return pspex, ckpt['editing_w'].clone().to(device)
    return pspex
```

If you want to use another vector, you need to train a new model on that vector.

**Comment by baopmessi** [comment]
I have a few questions:
- the ckpt in styleganex_edit_age.pt  is a model trained as fine-tuned from **Video face editing**. [https://github.com/williamyang1991/StyleGANEX#video-editing-1](url)
- The results of editing the face (young face) from the **Video face editing** are much better than **inversion + adding editing vectors to w+**  ?. Can i use inversion + editing vectors instead of training for each editing vectors

**Comment by williamyang1991** [comment]
- **inversion + editing vectors**: for image editing
- **training for each editing vectors**: for video editing, because in this way, we can use skip connection and temporal consistency loss to improve the quality and coherence of the video. And you will need no optimization per frame for inverion, which saves time during inferene.

**Comment by baopmessi** [comment]
Thanks you for your help!
 I have succeeded in inversion + editing vector (from the ckpt in your pretrained_models/styleganex_edit_age.pt). 
![image](https://github.com/williamyang1991/StyleGANEX/assets/34626618/699670c5-717f-46d0-b13d-bce742f59305)
However, I couldn't find where the editing vector comes from. I tried using editing vectors from the InterfaceGAN repo as #17  , but they didn't work. Could you please share how you obtained the editing vector as in pretrained_models/styleganex_edit_age.pt?
![image](https://github.com/williamyang1991/StyleGANEX/assets/34626618/959b6c86-8987-4f48-9ccc-3ac2d894621a)

**Comment by williamyang1991** [comment]
This is how I obtain the editing vector https://github.com/williamyang1991/StyleGANEX/issues/17#issuecomment-1797301586
I have no idea why your case fails.

In [huggingface](https://huggingface.co/spaces/PKUWilliamYang/StyleGANEX), we also provide the [editing vectors](https://huggingface.co/PKUWilliamYang/StyleGANEX/blob/main/direction_dics.pt)

```
self.editing_dicts = torch.load(hf_hub_download('PKUWilliamYang/StyleGANEX', 'direction_dics.pt'))
editing_w = self.editing_dicts['smile'].to(self.device)
```

**Comment by baopmessi** [comment]
Oh! I see.
- `self.editing_dicts = torch.load(hf_hub_download('PKUWilliamYang/StyleGANEX', 'direction_dics.pt'))
editing_w = self.editing_dicts['smile'].to(self.device)`
The smile direction in this code snippet is same to that in the LowRankGAN repo [link](https://github.com/zhujiapeng/LowRankGAN/blob/master/directions/ffhq1024/expression.npy), not in InterfaceGAN. 

- However, I couldn't find the age direction in the LowRankGAN repo, and the age vectors in InterfaceGAN don't match the "reduce_age" in your edit_direction. Can you help me obtain it?  Did you select the vector after running [link](https://github.com/zhujiapeng/LowRankGAN/tree/master#step-3-verify-the-directions-through-image-manipulation) ?  

Big thanks!

**Comment by williamyang1991** [comment]
Your interfacegan vector is based on StyleGAN-CelebHQ
You should check StyleGAN-FFHQ

**Comment by baopmessi** [comment]
I have tried all the age vectors in [1](https://github.com/genforce/interfacegan/tree/master/boundaries), but none of them work as well as yours. Because I want to retrain with some new expressions, I need to find the origin of the edit vectors.

**Comment by williamyang1991** [comment]
![image](https://github.com/williamyang1991/StyleGANEX/assets/18130694/1310d0af-d9b8-4eb3-87a5-c5f59c060b97)

I check my code. I use the edting vector from [anycostgan](https://github.com/mit-han-lab/anycost-gan/tree/master)

**Comment by nlcefn** [comment]
Can I edit an image using the StyleGANEX inversion model, similar to an image-to-image model, by editing the vector to achieve multiple styles, such as glasses? Meanwhile, when using the StyleGANEX model to transform images, can I edit the picture with one model for multiple styles by calculating the editing vector?

**Comment by nlcefn** [comment]
Can I edit an image using the StyleGANEX inversion model, similar to an image-to-image model with skip connections, by modifying the vector to achieve various styles, such as adding glasses? Additionally, when using the StyleGANEX model to transform images, can I edit the picture for multiple styles using a single model by calculating the editing vector, and does this process involve skip connections?

**Comment by williamyang1991** [comment]
Two ways of editing
1. StyleGANEX inversion + arbitrary editing vector. 
2. Train encoder with skip connections, to build an image-to-image model on the editing vector of adding glasses. Once trained, this encoder with skip connections will only work on this editing vector. Not work for other editing vectors.

Both way can handle arbitrary face images.

**Comment by nlcefn** [comment]
If I want to train a model with skip connections for another editing style, how should I start the training process, and how many datasets should I prepare?and how should I specify the command-line parameters for training?
